packages/terra-dropdown-button/src/DropdownButton.jsx

```diff
@@ -57,29 +57,56 @@ class DropdownButton extends React.Component {
     this.handleDropdownRequestClose = this.handleDropdownRequestClose.bind(this);
     this.handleKeyDown = this.handleKeyDown.bind(this);
     this.handleKeyUp = this.handleKeyUp.bind(this);
+    this.setButtonNode = this.setButtonNode.bind(this);
+    this.getButtonNode = this.getButtonNode.bind(this);
+    this.setListNode = this.setListNode.bind(this);
+    this.state = { isOpen: false, isActive: false, openedViaKeyboard: false };
+  }
+
+  setListNode(element) {
+    this.dropdownList = element;
+  }
 
-    this.state = { isOpen: false, isActive: false };
+  setButtonNode(node) {
+    this.buttonNode = node;
+  }
+
+  getButtonNode() {
+    return this.buttonNode;
   }
 
   handleDropdownButtonClick() {
+    if (this.state.isOpen) {
+      this.setState({ openedViaKeyboard: false });
+    }
     this.setState(prevState => ({ isOpen: !prevState.isOpen }));
   }
 
   handleDropdownRequestClose(callback) {
-    this.setState({ isOpen: false }, typeof callback === 'function' ? callback : undefined);
+    const callbackOnRequestClose = typeof callback === 'function' ? callback : undefined;
+    this.setState({ isOpen: false, openedViaKeyboard: false, isActive: false }, callbackOnRequestClose);
   }
 
   /*
     In FireFox active styles don't get applied on space
    */
   handleKeyDown(event) {
-    if (event.keyCode === KeyCode.KEY_SPACE) {
+    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
       this.setState({ isActive: true });
+      this.setState({ openedViaKeyboard: true });
+    } else if (event.keyCode === KeyCode.KEY_DOWN && this.state.isOpen && !this.state.openedViaKeyboard) {
```

Comment:
Is `isOpen` check merely a safety check here?

Reply:
yes, to ensure querySelector() is triggered only when dropdown is open.